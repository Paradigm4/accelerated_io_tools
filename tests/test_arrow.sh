#!/bin/bash

set -o errexit

DIR=`dirname $0`
TEST_OUT=$DIR/test_arrow.out
IQ="iquery --afl --query"
F=/tmp/1

function iq() {
    $IQ "$1" >> $TEST_OUT
}


: > $TEST_OUT                   # Reset output file

iq "aio_save(apply(build(<x:int64>[i=-10:10], i), y, iif(i%2=0, i, int64(null))), '$F', 'format=arrow')"
python -c "import pyarrow; print(pyarrow.open_stream('$F').read_all().to_pandas())" \
    >> $TEST_OUT

iq "aio_save(apply(build(<x:int64>[i=1:100:0:20], i), y, iif(i%2=0, i, int64(null))), '$F', 'format=arrow')"
python -c "import pyarrow; print(pyarrow.open_stream('$F').read_all().to_pandas().sort_values('x').to_string(index=False))" \
    >> $TEST_OUT

iq "aio_save(apply(build(<x:int64>[i=1:20:0:4], i), y, iif(i%2=0, string(i), string(null))), '$F', 'format=arrow')"
python -c "import pyarrow; print(pyarrow.open_stream('$F').read_all().to_pandas().sort_values('x').to_string(index=False))" \
    >> $TEST_OUT

$IQ "aio_save(build(<x:uint64>[i=0:0], i), '$F', 'format=arrow')" 2>&1 \
    |  sed --expression='s/ line: [0-9]\+//g'                          \
    |  grep --invert-match "Failed query id:" >> $TEST_OUT             \
    || echo "expected exception"

$IQ "aio_save(build(<x:int64>[i=0:0], i), 'console', 'format=arrow')" \
    >> $TEST_OUT
$IQ "aio_save(build(<x:int64>[i=0:0], i), 'stdout', 'format=arrow')" \
    >> $TEST_OUT
$IQ "aio_save(build(<x:int64>[i=0:0], i), 'stderr', 'format=arrow')" \
    >> $TEST_OUT


echo "I. int64, int64(null)"
echo "1. store"
time $IQ "set no fetch; store(apply(build(<x:int64 not null>[i=1:3100000], i), y, iif(i%2=0, i * i, int64(null))), foo)" \
     >> $TEST_OUT
iq "summarize(foo)"

echo "2. save"
time $IQ "save(foo, '$F', 0, '(int64, int64 null)')" \
     >> $TEST_OUT
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)" \
     >> $TEST_OUT

echo "3. aio_save(binary)"
time $IQ "aio_save(foo, '$F', 'format=(int64, int64 null)')" \
     >> $TEST_OUT
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)" \
     >> $TEST_OUT

echo "4. aio_save(arrow)"
time $IQ "aio_save(foo, '$F', 'format=arrow')" \
     >> $TEST_OUT
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)" \
     >> $TEST_OUT

echo "5. Arrow read"
time python -c "import pyarrow; print(len(pyarrow.open_stream('$F').read_pandas()))" \
     >> $TEST_OUT

echo "6. SciDB-Py fetch"
time python -c "import scidbpy; print(len(scidbpy.connect().arrays.foo.fetch(atts_only=True)))" \
     >> $TEST_OUT

echo "7. SciDB-Py fetch w/ Arrow"
time python -c "import scidbpy; print(len(scidbpy.connect().arrays.foo.fetch(atts_only=True, use_arrow=True)))" \
     >> $TEST_OUT

iq "remove(foo)"


echo "II. int64, string(null)"
echo "1. store"
time $IQ "set no fetch; store(apply(build(<x:int64 not null>[i=1:2700000], i), y, iif(i%2=0, string(i * i), string(null))), foo)" \
     >> $TEST_OUT
iq "summarize(foo)"

echo "2. save"
time $IQ "save(foo, '$F', 0, '(int64, string null)')" \
     >> $TEST_OUT
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)" \
     >> $TEST_OUT

echo "3. aio_save(binary)"
time $IQ "aio_save(foo, '$F', 'format=(int64, string null)')" \
     >> $TEST_OUT
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)" \
     >> $TEST_OUT

echo "4. aio_save(arrow)"
time $IQ "aio_save(foo, '$F', 'format=arrow')" \
     >> $TEST_OUT
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)" \
     >> $TEST_OUT

echo "5. Arrow read"
time python -c "import pyarrow; print(len(pyarrow.open_stream('$F').read_pandas()))" \
     >> $TEST_OUT

echo "6. SciDB-Py fetch"
time python -c "import scidbpy; print(len(scidbpy.connect().arrays.foo.fetch(atts_only=True)))" \
     >> $TEST_OUT

echo "7. SciDB-Py fetch w/ Arrow"
time python -c "import scidbpy; print(len(scidbpy.connect().arrays.foo.fetch(atts_only=True, use_arrow=True)))" \
     >> $TEST_OUT


iq "remove(foo)"


diff $TEST_OUT $DIR/test_arrow.expected
