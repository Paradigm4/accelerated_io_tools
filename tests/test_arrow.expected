{chunk_no,dest_instance_id,source_instance_id} val
     x     y
0  -10 -10.0
1   -9   NaN
2   -8  -8.0
3   -7   NaN
4   -6  -6.0
5   -5   NaN
6   -4  -4.0
7   -3   NaN
8   -2  -2.0
9   -1   NaN
10   0   0.0
11   1   NaN
12   2   2.0
13   3   NaN
14   4   4.0
15   5   NaN
16   6   6.0
17   7   NaN
18   8   8.0
19   9   NaN
20  10  10.0
{chunk_no,dest_instance_id,source_instance_id} val
x      y
  1    NaN
  2    2.0
  3    NaN
  4    4.0
  5    NaN
  6    6.0
  7    NaN
  8    8.0
  9    NaN
 10   10.0
 11    NaN
 12   12.0
 13    NaN
 14   14.0
 15    NaN
 16   16.0
 17    NaN
 18   18.0
 19    NaN
 20   20.0
 21    NaN
 22   22.0
 23    NaN
 24   24.0
 25    NaN
 26   26.0
 27    NaN
 28   28.0
 29    NaN
 30   30.0
 31    NaN
 32   32.0
 33    NaN
 34   34.0
 35    NaN
 36   36.0
 37    NaN
 38   38.0
 39    NaN
 40   40.0
 41    NaN
 42   42.0
 43    NaN
 44   44.0
 45    NaN
 46   46.0
 47    NaN
 48   48.0
 49    NaN
 50   50.0
 51    NaN
 52   52.0
 53    NaN
 54   54.0
 55    NaN
 56   56.0
 57    NaN
 58   58.0
 59    NaN
 60   60.0
 61    NaN
 62   62.0
 63    NaN
 64   64.0
 65    NaN
 66   66.0
 67    NaN
 68   68.0
 69    NaN
 70   70.0
 71    NaN
 72   72.0
 73    NaN
 74   74.0
 75    NaN
 76   76.0
 77    NaN
 78   78.0
 79    NaN
 80   80.0
 81    NaN
 82   82.0
 83    NaN
 84   84.0
 85    NaN
 86   86.0
 87    NaN
 88   88.0
 89    NaN
 90   90.0
 91    NaN
 92   92.0
 93    NaN
 94   94.0
 95    NaN
 96   96.0
 97    NaN
 98   98.0
 99    NaN
100  100.0
{chunk_no,dest_instance_id,source_instance_id} val
x     y
 1  None
 2     2
 3  None
 4     4
 5  None
 6     6
 7  None
 8     8
 9  None
10    10
11  None
12    12
13  None
14    14
15  None
16    16
17  None
18    18
19  None
20    20
UserException in file: PhysicalAioSave.cpp function: attributes2ArrowSchema
Error id: scidb::SCIDB_SE_ARRAY_WRITER::SCIDB_LE_ILLEGAL_OPERATION
Error description: Error while saving array. Illegal operation: Type uint64 not supported in arrow format.
{chunk_no,dest_instance_id,source_instance_id} val
{chunk_no,dest_instance_id,source_instance_id} val
{chunk_no,dest_instance_id,source_instance_id} val
Query was executed successfully
{inst,attid} att,count,bytes,chunks,min_count,avg_count,max_count,min_bytes,avg_bytes,max_bytes
{0,0} 'all',3100000,74400720,12,100000,775000,1000000,48,6.20006e+06,16000060
{i} x,y
50 MB (52700000 B)
{chunk_no,dest_instance_id,source_instance_id} val
50 MB (52700000 B)
{chunk_no,dest_instance_id,source_instance_id} val
47 MB (49988260 B)
3100000
3100000
3100000
Query was executed successfully
Query was executed successfully
{inst,attid} att,count,bytes,chunks,min_count,avg_count,max_count,min_bytes,avg_bytes,max_bytes
{0,0} 'all',2700000,78919312,9,700000,900000,1000000,48,8.76881e+06,21500060
{i} x,y
50 MB (53268772 B)
{chunk_no,dest_instance_id,source_instance_id} val
50 MB (53268772 B)
{chunk_no,dest_instance_id,source_instance_id} val
47 MB (49556892 B)
2700000
2700000
2700000
Query was executed successfully
